File: backend/src/utils/top_posts_parser.py
"""
Parser for TOP POSTS sheet from LinkedIn Excel export.
Extracts posts with engagement metrics and URLs.
"""

from typing import Dict, Any, List, Optional
from openpyxl import load_workbook
import io
from datetime import datetime


def parse_top_posts_sheet(file_content: bytes) -> List[Dict[str, Any]]:
    """
    Parse the TOP POSTS sheet from LinkedIn Excel export.

    Expected structure:
    Row 1: Header note
    Row 2: Empty
    Row 3: Column headers - "Post URL", "Post publish date", "Engagements", ..., "Post URL", "Post publish date", "Impressions"
    Row 4+: Data rows with post information

    The sheet has two views of the data:
    - Left side (columns A-C): Post URL, Publish Date, Engagements
    - Right side (columns E-G): Post URL, Publish Date, Impressions

    Args:
        file_content: Binary content of Excel file

    Returns:
        List of dictionaries containing post data with keys:
        - url: str
        - publish_date: str
        - engagements: float
        - impressions: float (optional)
    """
    try:
        wb = load_workbook(io.BytesIO(file_content), data_only=True)

        if "TOP POSTS" not in wb.sheetnames:
            raise ValueError("TOP POSTS sheet not found in Excel file")

        ws = wb["TOP POSTS"]

        posts = {}  # Use dict to merge engagement and impression data by URL

        # Row 3 contains headers, data starts from row 4
        # Parse engagement data from columns A-C
        for row_idx in range(4, ws.max_row + 1):
            url = ws.cell(row_idx, 1).value
            publish_date = ws.cell(row_idx, 2).value
            engagements = ws.cell(row_idx, 3).value

            # Stop if we hit empty rows
            if url is None or publish_date is None or engagements is None:
                # Check if all first 3 columns are empty (end of data)
                if (ws.cell(row_idx, 1).value is None and
                    ws.cell(row_idx, 2).value is None and
                    ws.cell(row_idx, 3).value is None):
                    break

            # Skip empty rows
            if url is None or publish_date is None:
                continue

            # Convert engagements to float
            try:
                engagements_count = float(engagements) if engagements else 0
            except (ValueError, TypeError):
                engagements_count = 0

            url_str = str(url)
            posts[url_str] = {
                "url": url_str,
                "publish_date": str(publish_date),
                "engagements": engagements_count,
                "impressions": 0
            }

        # Parse impression data from columns E and G and merge with engagement data
        # In openpyxl, columns are 1-indexed: A=1, B=2, C=3, D=4, E=5, F=6, G=7
        for row_idx in range(4, ws.max_row + 1):
            url = ws.cell(row_idx, 5).value  # Column E
            impressions = ws.cell(row_idx, 7).value  # Column G

            if url is None or impressions is None:
                if (ws.cell(row_idx, 5).value is None and
                    ws.cell(row_idx, 7).value is None):
                    break
                continue

            # Convert impressions to float
            try:
                impressions_count = float(impressions) if impressions else 0
            except (ValueError, TypeError):
                impressions_count = 0

            url_str = str(url)
            if url_str in posts:
                posts[url_str]["impressions"] = impressions_count
            else:
                # If URL not found in engagement data, still add it
                posts[url_str] = {
                    "url": url_str,
                    "publish_date": "",
                    "engagements": 0,
                    "impressions": impressions_count
                }

        # Convert dict to sorted list
        posts_list = list(posts.values())

        # Sort by engagements descending
        posts_list.sort(key=lambda x: x["engagements"], reverse=True)

        return posts_list

    except Exception as e:
        raise ValueError(f"Failed to parse TOP POSTS sheet: {str(e)}")

def get_top_posts(posts: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Get the top 6 posts by engagement count.

    Args:
        posts: List of post dictionaries

    Returns:
        Top 6 posts with rank added
    """
    top_posts = posts[:6]
    for idx, post in enumerate(top_posts, 1):
        post["rank"] = idx
    return top_posts

File: site/src/utils/excel/topPostsParser.ts
/**
 * Top posts Parser
 * ================
 * Parses the "Top posts" sheet from LinkedIn analytics Excel export.
 *
 * The Top posts sheet contains individual post performance metrics:
 * - Post URL (unique identifier)
 * - Publish date
 * - Engagement count
 * - Impression count
 *
 * Expected Excel Structure:
 * - Sheet name: "Top posts"
 * - Two-column layout: left side (engagements) and right side (impressions)
 * - Left: Columns A=URL, B=Date, C=Engagements
 * - Right: Columns E=URL, F=Date, G=Impressions
 * - Data rows start after header row
 *
 * Performance Notes:
 * - Uses Map for O(1) URL deduplication
 * - Single pass through rows
 * - ~5-50ms processing time depending on post count
 * - Memory efficient: only stores unique URLs
 */
import type { WorkBook } from 'xlsx';
import type { TopPost } from '@types';
import { 
  parseDate, 
  parseNumber, 
  parseURL, 
  findSheet,
  findRowWithKeywords,
  iterateRowsUntilEmpty,
  getRowValues
} from '@utils/excel/utils';
import { SHEET_NAMES, COLUMN_LAYOUTS, KEYWORDS, MAX_ROWS } from '@utils/excel/constants';

interface PostData {
  url: string;
  date: string;
  engagements: number;
  impressions: number;
}

/**
 * Type for specifying which metric type is being processed
 */
type MetricType = 'engagements' | 'impressions';

/**
 * Column configuration for a metric side (left or right)
 */
interface MetricColumns {
  URL: string;
  DATE: string;
  VALUE: string;
}

/**
 * Process a single post entry and update the posts map
 * @param postsMap - Map storing all posts indexed by URL
 * @param columns - Column configuration for this metric type
 * @param rowValues - Values from the current row
 * @param metricType - Type of metric being processed
 */
function processPostEntry(
  postsMap: Map<string, PostData>,
  columns: MetricColumns,
  rowValues: Record<string, any>,
  metricType: MetricType
): void {
  const url = parseURL(rowValues[columns.URL]);
  if (!url) return;

  const date = parseDate(rowValues[columns.DATE]);
  const metric = parseNumber(rowValues[columns.VALUE]);

  // Get or create post entry
  if (!postsMap.has(url)) {
    postsMap.set(url, {
      url,
      date,
      engagements: 0,
      impressions: 0,
    });
  }

  // Update post with new metric (use max value if duplicate)
  const post = postsMap.get(url)!;
  post[metricType] = Math.max(post[metricType], metric);
  
  // Update date if not already set
  if (!post.date && date) {
    post.date = date;
  }
}

/**
 * Parse the Top posts sheet to extract top performing posts
 * @param workbook - Parsed Excel workbook from xlsx library
 * @returns Array of TopPost objects in the order they appear in the spreadsheet (already sorted by engagement)
 */
export function parseTopPosts(workbook: WorkBook): TopPost[] {
  const sheet = findSheet(workbook, SHEET_NAMES.TOP_POSTS);
  if (!sheet) {
    console.warn('Top posts sheet not found');
    return [];
  }

  try {
    // Use Map to deduplicate posts by URL
    const postsMap = new Map<string, PostData>();

    // Find header row
    const headerRow = findRowWithKeywords(sheet, ['A', 'B'], KEYWORDS.HEADERS.URL);
    if (!headerRow) {
      console.warn('Could not find header row in top posts sheet');
      return [];
    }

    const layout = COLUMN_LAYOUTS.TOP_POSTS;
    const allColumns = [
      layout.ENGAGEMENTS.URL, 
      layout.ENGAGEMENTS.DATE, 
      layout.ENGAGEMENTS.VALUE,
      layout.IMPRESSIONS.URL, 
      layout.IMPRESSIONS.DATE, 
      layout.IMPRESSIONS.VALUE
    ];

    // Process data rows using iterator
    for (const row of iterateRowsUntilEmpty(sheet, headerRow + 1, allColumns, MAX_ROWS.DATA)) {
      const values = getRowValues(sheet, row, allColumns);

      // Process left side (Engagements)
      processPostEntry(postsMap, layout.ENGAGEMENTS, values, 'engagements');

      // Process right side (Impressions)
      processPostEntry(postsMap, layout.IMPRESSIONS, values, 'impressions');
    }

    // Convert map to array with ranks
    return Array.from(postsMap.values()).map((post, index) => ({
      rank: index + 1,
      url: post.url,
      publish_date: post.date,
      engagements: post.engagements,
      impressions: post.impressions,
    }));
  } catch (error) {
    console.error('Error parsing Top posts sheet:', error);
    return [];
  }
}
